modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java

```diff
@@ -1827,6 +1762,94 @@ private CompletableFuture<Object> processMultiEntryAction(ReadWriteMultiRowRepli
         }
     }
 
+    /**
+     * Precesses multi request.
+     *
+     * @param request Multi request operation.
+     * @param txCoordinatorId Transaction coordinator id.
+     * @return Listener response.
+     */
+    private CompletableFuture<Object> processMultiEntryAction(ReadWriteMultiRowPkReplicaRequest request, String txCoordinatorId) {
+        UUID txId = request.transactionId();
+        TablePartitionId committedPartitionId = request.commitPartitionId();
+        List<BinaryTuple> primaryKeys = request.primaryKeys();
+
+        assert committedPartitionId != null || request.requestType() == RequestType.RW_GET_ALL
+                : "Commit partition is null [type=" + request.requestType() + ']';
+
+        switch (request.requestType()) {
+            case RW_GET_ALL: {
+                CompletableFuture<BinaryRow>[] rowFuts = new CompletableFuture[primaryKeys.size()];
+
+                int i = 0;
+
+                for (BinaryTuple primaryKey : primaryKeys) {
+                    rowFuts[i++] = resolveRowByPk(primaryKey, txId, (rowId, row) -> {
+                        if (rowId == null) {
+                            return completedFuture(null);
+                        }
+
+                        return takeLocksForGet(rowId, txId)
+                                .thenApply(ignored -> row);
+                    });
+                }
+
+                return allOf(rowFuts)
+                        .thenApply(ignored -> {
+                            var result = new ArrayList<BinaryRow>(primaryKeys.size());
+
+                            for (CompletableFuture<BinaryRow> rowFut : rowFuts) {
+                                result.add(rowFut.join());
+                            }
+
+                            return result;
+                        });
+            }
+            case RW_DELETE_ALL: {
+                CompletableFuture<RowId>[] rowIdLockFuts = new CompletableFuture[primaryKeys.size()];
+
+                for (int i = 0; i < primaryKeys.size(); i++) {
+                    rowIdLockFuts[i] = resolveRowByPk(primaryKeys.get(i), txId, (rowId, row) -> {
+                        if (rowId == null) {
+                            return completedFuture(null);
+                        }
```

Comment:
This is not related to your changes, I guess, but let's examine the following situation (table is empty at the start):
```
    |   TX 1   |   TX 2   |
ts0 |   start  |   start  |
ts1 | delete A | delete B |
ts2 | insert B | insert A |
ts3 |  commit  | commit   |
```
In serializable transactions, either A or B will be missing after the execution. But, according to this code, we will ignore locks on delete, resulting in both A and B being inserted by corresponding transactions. It breaks the protocol. We should consult owners of the feature.
My guess - taking locks based on RowId is a wrong idea. We should take locks based exclusively on PK. We don't even need to resolve it to do so. What do you think?

Reply:
@sanpwc Can you please help me to resolve this comment?

Reply:
Not sure that I got you guys. We do acquire locks on PK while index "scan"
```
    private <T> CompletableFuture<T> resolveRowByPk(
            BinaryTuple pk,
            UUID txId,
            BiFunction<@Nullable RowId, @Nullable BinaryRow, CompletableFuture<T>> action
    ) {
        IndexLocker pkLocker = indexesLockers.get().get(pkIndexStorage.get().id());

        assert pkLocker != null;

        return pkLocker.locksForLookupByKey(txId, pk)
```

Reply:
I didn't realize that "resolveRowByPk" already has a lock. If we already do that, what's the reason of having locks for RowIds?

Reply:
Quick answer because theoretical lock model requires it. E.g. you may use different indexes as an entry points for the rows.